scripts/evaluate_reconstruction.py

```diff
@@ -0,0 +1,259 @@
+elif data_name == 'paul':
+  X, y, encoder = get_paul(
+    'data/paul15/house_keeping_genes_Mouse_bone_marrow.txt',
+    'data/paul15/house_keeping_genes_Mouse_HSC.txt',
+  )
+elif data_name == 'cite_seq':
+  X, y, encoder = get_citeseq('data/cite_seq/CITEseq.h5ad')
+elif data_name == 'mouse_brain':
+  X, y, encoder = get_mouse_brain(
+    'data/mouse_brain_broad/mouse_brain_all_cells_20200625.h5ad',
+    'data/mouse_brain_broad/snRNA_annotation_astro_subtypes_refined59_20200823.csv',
+    log_transform=False, #scVI requires counts, so we will normalize and log transform after.
+  )
+
+  scVI_X = X
+
+  marker_map_adata = anndata.AnnData(X.copy())
+  sc.pp.normalize_per_cell(marker_map_adata, counts_per_cell_after=1e4)
+  sc.pp.log1p(marker_map_adata)
+  sc.pp.scale(marker_map_adata, max_value=10)
+  X = marker_map_adata.X.copy()
+
+result_options = [
+  'l2_all',
+  'l2_groups',
+  'jaccard_all',
+  'jaccard_groups',
+  'spearman_rho_all',
+  'spearman_rho_groups',
+  'spearman_p_all',
+  'spearman_p_groups',
+]
+model_options = ['marker_map', 'scVI']
+results = { 'marker_map': { z: None for z in result_options}, 'scVI': { z: None for z in result_options} }
+
+for i in range(num_times):
+  # we will use 80% training data, 20% validation data during the training of the marker map
+  train_dataloader, val_dataloader, _, train_indices, val_indices, test_indices = split_data_into_dataloaders(
+      X,
+      y,
+      train_size=0.7,
+      val_size=0.1,
+  )
+  y_test = y[test_indices]
+  groups = np.unique(y)
+
+  for model in model_options:
+    if model == 'marker_map':
+      model_X = X
+      X_test = model_X[test_indices, :]
+
+      recon_X_test = trainAndGetReconMarkerMap(
+        hidden_layer_size,
+        z_size,
+        k,
+        train_dataloader,
+        val_dataloader,
+        X_test,
+        gpus,
+      )
+
+    if model == 'scVI':
+      model_X = scVI_X
+      X_test = model_X[test_indices, :]
+      recon_X_test = trainAndGetRecon_scVI(
+        model_X[np.concatenate([train_indices, val_indices]), :],
+        X_test,
+      )
+
+    l2_all, l2_by_group = getL2(X_test, recon_X_test, y_test, groups)
+    jaccard_all, jaccard_index, spearman_rho_all, spearman_rho, spearman_p_all, spearman_p = analyzeVariance(
+      X_test,
+      recon_X_test,
+      y_test,
+      groups,
+      int(X_test.shape[1]*0.2),
+    )
+
+    # this is really trash code
+    insertOrConcatenate(results, model, 'l2_all', l2_all)
+    insertOrConcatenate(results, model, 'l2_groups', l2_by_group)
+    insertOrConcatenate(results, model, 'jaccard_all', jaccard_all)
+    insertOrConcatenate(results, model, 'jaccard_groups', jaccard_index)
+    insertOrConcatenate(results, model, 'spearman_rho_all', spearman_rho_all)
+    insertOrConcatenate(results, model, 'spearman_rho_groups', spearman_rho)
+    insertOrConcatenate(results, model, 'spearman_p_all', spearman_p_all)
+    insertOrConcatenate(results, model, 'spearman_p_groups', spearman_p)
```

Comment:
Let's mark it as an "issue" when the PR is accepted. I don't think it is a problem. It is pretty clear what it is doing, but ye there is probably a cleaner way to do it.